python/pyspark/ml/stat/distribution.py

```diff
@@ -0,0 +1,267 @@
+        return -0.5 * (dim * np.log(2 * np.pi) + self.log_det_cov + maha)
+
+
+     
+
+    def __process_parameters(self, dim, mean, cov):
+        """
+        Helper funtion to process input values, based on scipy multivariate
+
+        Infer dimensionality from mean or covariance matrix, ensure that
+        mean and covariance are full vector resp. matrix.
+
+        """
+
+        # Try to infer dimensionality
+        if dim is None:
+            if mean is None:
+                if cov is None:
+                    dim = 1
+                else:
+                    cov = np.asarray(cov, dtype=float)
+                    if cov.ndim < 2:
+                        dim = 1
+                    else:
+                        dim = cov.shape[0]
+            else:
+                mean = np.asarray(mean, dtype=float)
+                dim = mean.size
+        else:
+            if not np.isscalar(dim):
+                raise ValueError("Dimension of random variable must be a scalar.")
+
+        # Check input sizes and return full arrays for mean and cov if necessary
+        if mean is None:
+            mean = np.zeros(dim)
+        mean = np.asarray(mean, dtype=float)
+
+        if cov is None:
+            cov = 1.0
+        cov = np.asarray(cov, dtype=float)
+
+        if dim == 1:
+            mean.shape = (1,)
+            cov.shape = (1, 1)
+
+        if mean.ndim != 1 or mean.shape[0] != dim:
+            raise ValueError("Array 'mean' must be vector of length %d." % dim)
+        if cov.ndim == 0:
+            cov = cov * np.eye(dim)
+        elif cov.ndim == 1:
+            cov = np.diag(cov)
+        else:
+            if cov.shape != (dim, dim):
+                raise ValueError("Array 'cov' must be at most two-dimensional,"
+                                     " but cov.ndim = %d" % cov.ndim)
+
+        return dim, mean, cov
+
+    def __process_quantiles(self, x, dim):
+        """
+        Helper funtion to process quantiles, based on scipy multivariate
+
+        Adjust quantiles array so that last axis labels the components of
+        each data point.
+        
+        """
+        x = np.asarray(x, dtype=float)
+
+        if x.ndim == 0:
+            x = x[np.newaxis]
+        elif x.ndim == 1:
+            if dim == 1:
+                x = x[:, np.newaxis]
+            else:
+                x = x[np.newaxis, :]
+
+        return x
+
+
+    def __pinv_1d(self, v, eps=1e-5):
+        """
+        A helper function for computing the pseudoinverse, based on scipy multivariate 
+
+        v : iterable of numbers
+            This may be thought of as a vector of eigenvalues or singular values.
+        eps : float
+            Elements of v smaller than eps are considered negligible.
+        returns  1d float ndarray
+            A vector of pseudo-inverted numbers. 
+        """
+        return np.array([0 if abs(x) < eps else 1/x for x in v], dtype=float)
```

Comment:
This can be easily vectorized. You are creating a copy anyway.

    v_c = v.copy()
    mask = v_c <= eps
    v_c[mask] = 0.0
    v_c[~mask] = 1. / v_c[~mask]